### DISCUSSION TITLE
LiTime BMS

### DISCUSSION THREAD

**Discussion by calledit** [Louisvdw/dbus-serialbattery#1098] [New BMS requests]
### Is your feature request related to a problem? Please describe.

Hi I built a project to request and interpret data from the LiTime BMS:

https://github.com/calledit/LiTime_BMS_bluetooth

I don't have that much experience with venus OS. But would be nice if the LiTime BMS was possible to use with venus OS.

This is basically a request for someone to port that project so it can be used with Venus OS.

 The LiTime_BMS_bluetooth system is super simple. There are more or less just 4 parts:
* Connect to the BMS using BLE.
* Send the command: commands.query_battery_status
* Wait for a notification from the BMS then calls the function interpret_data()
* Sleep a while then send the query_battery_status command again and repeat.

### Describe the solution you'd like

The solution I would like is if someone with experience with Venus OS and this project could initiate this. It would be nice if someone did the boiler plate stuff. The project does have a bluetooth low energy BMS already tried to look at that but it was a bit more complicated compared to how the LiTime protocol works. 

### Describe alternatives you've considered

Possible alterantive is that I do it myself from the start but it would take much longer for me since I would need to learn how this project is setup.

**Comment by mr-manuel** [comment]
See https://mr-manuel.github.io/venus-os_dbus-serialbattery/general/supported-bms#addrequest-new-bms.

**Reply by calledit** [reply]
Okay so I started working on a PR.
https://github.com/calledit/dbus-serialbattery/blob/patch-1

But some questions:

When looking at the other BMSes that use bluetooth we have:
* jkbms_ble.py which is actually only a partial implementation as some of the Bluetooth specific code is located in jkbms_brn.py
* lltjbd_ble.py looks like a cleaner implementation.

In both of them we have code to kill various bluetooth processes. Do I need to copy that? Is the bluetooth on Venus OS very unstable?


Here is how it looks like bluetooth is solved  in lltjbd_ble.py.
BatteryClass.__init__() is executed
   threading.Thread starts a new thread running background_loop() spins in a while loop running bt_main_loop() over and over
     bt_main_loop() trys to connect to the BMS then goes in to a loop where it sleeps until the connection fails.

This is the only thing lltjbd_ble.py does except that it also implements read_serial_data_llt() which is called from lltjbd.py to actually parse and request data from the BMS.

This makes sense for lltjbd which also have a serial interface. Don't know if it makes sense to build a LiTime class in the same way due to the fact that it does not have that serial interface, but maybe. I don't know, any opinions?

**Reply by mr-manuel** [reply]
Please make a fork of my repository and open the PR there, this repository is outdated.

The implementation of the `jkbms_brn.py` was thought to be used also with other projects, while `jkbms_ble.py` contains the modifications that are needed for this driver.

If I compare the JKBMS BLE and JBD/LLT BLE implementation, than there are more issues with the JKBMS BLE implementation than with the JBD/LLT BLE implementation. But could also be that JBD/LLT BLE is not used that much.

If the LiTime BMS only has a BLE connection then make only one script file. If it has serial communication and BLE communication with the same communication protocol, then you can make it the same way as the LLT/JBD driver.

Are you familiar with Python and BLE connection/communication? Since this drivers were created by try and error, therefore for sure they don't follow a best practice.

**Reply by calledit** [reply]
Okay would be good if it is written that it is outdated in the README of this project.

Quite familiar with BLE, and python but not in combination.

So I am now trying to figure out how to report back the values that the BMS Is giving.

Seams like you implement refresh_data() that sets a bunch of values. The values as listed in the function battery.py > init_values() it is unclear to me which values should be filled and which are calculated from other values ie self.current_avg. Then we have other values where the unit is unclear self.capacity_remain is it amps or watt or SOC?

Basically I am wondering what variables should be filled, what units are they in and it would also be good to know what they are used for.

We can take an example where the LiTime BMS reports a value for current. But that value is supper inaccurate for some reason; it fluctuates wildly each time you read it, sometimes it is +10 amps, the next reading it might be 0 amps, the next +2 amps and the next -2 amps. Despite the load being super stable. The only way to get an accurate reading for current is by reading another value "remaining Ah" and read that multiple times then divide with the time between reading.

Since I don't know what the use for the self.current value is I don't know what to repport; the unstable one reported by the BMS or the accurate one.

Here are all the values that the BMS reports: What maps to what?
```
{
'messured_total_voltage'
'cells_added_together_voltage'
'mosfet_temp'
'remaining_amph'
'full_charge_capacity_amph'
'protection_state'//Non all zeros means the BMS is "protecting" the battery somehow. (00000004 = Over Charge Protection, 00000020 = Over-discharge Protection, 00000040 = Charging Over Current Protection, 00000080 = Discharging Over Current Protection, 00000100 = High-temp Protection, 00000200 = High-temp Protection, 00000400 = Low-temp Protection, 00000800 = Low-temp Protection, 00004000 = Short Circuit Protection)
'heat',//discharge disabled due to app button = 00000080, heater_error = 00000002
'failure_state'//First numer/byte of the three regards cell_error, second number is also cell_error, third number is BMS_error
'is_balancing'//("1" or "0")
'battery_state'://charge disabled = "0004", charging = "0001" (when charging active app will show estimated time untill fully charged), discharging/idle: "0000", unkown = "0002"
'SOC'
'SOH'
'discharges_count'
'discharges_amph_count'
'cell_temp'
'current'
cellsVoltages' // array of voltages
}
```

**Reply by mr-manuel** [reply]
Did you check the [template](https://github.com/mr-manuel/venus-os_dbus-serialbattery/blob/master/etc/dbus-serialbattery/bms/battery_template.py)? If yes and you did not understand it, I will update it to be more clear.

Best would be, if you take the template and start from there 

Capacity is Ah, SOC is %.

**Reply by calledit** [reply]
Yes I checked the template, I understand it but it lacks many variables.

The template only contains 3 variables: (if you exclude the settings)
```
voltage, current, soc = unpack_from(">hxxhh", soc_data)
        self.voltage = voltage / 10
        self.current = current / -10
        self.soc = soc / 10
```

And as I stated I have the 16+ variables that I listed. Without proper descriptions of each variable it will be hard to know what variable does what.

The battery.py file has more variables (some I understand should be filled out some should not) the best place I could find with verbal descriptions of the variables is the feature page. But you sort of have to guess what feature correspond to what variable. https://mr-manuel.github.io/venus-os_dbus-serialbattery/general/features/

Variables as listed in the battery.py file.
```
self.voltage: float = None
        self.current: float = None
        self.current_avg: float = None
        self.current_avg_lst: list = []
        self.current_corrected: float = None
        self.current_external: float = None
        self.capacity_remain: float = None
        self.capacity: float = None
        self.cycles: float = None
        self.total_ah_drawn: float = None
        self.production = None
        self.protection = Protection()
        self.version = None
        self.time_to_soc_update: int = 0
        self.temp_sensors: int = None
        self.temp1: float = None
        self.temp2: float = None
        self.temp3: float = None
        self.temp4: float = None
        self.temp_mos: float = None
        self.cells: List[Cell] = []
        self.control_voltage: float = None
        self.soc_reset_requested: bool = False
        self.soc_reset_last_reached: int = 0  # save state to preserve on restart
        self.soc_reset_battery_voltage: int = None
        self.max_battery_voltage: float = None
        self.min_battery_voltage: float = None
        self.allow_max_voltage: bool = True  # save state to preserve on restart
        self.max_voltage_start_time: int = None  # save state to preserve on restart
        self.transition_start_time: int = None
        self.charge_mode: str = None
        self.charge_mode_debug: str = ""
        self.charge_mode_debug_float: str = ""
        self.charge_mode_debug_bulk: str = ""
        self.charge_limitation: str = None
        self.discharge_limitation: str = None
        self.linear_cvl_last_set: int = 0
        self.linear_ccl_last_set: int = 0
        self.linear_dcl_last_set: int = 0
        # list of available callbacks, in order to display the buttons in the GUI
        self.available_callbacks: List[str] = []
```

Would be great if the template was expanded so that it documents all variables that could/should be filled out. Which ones are required which are not and so on. If they are expected to be floats or strings and so on. It would make things allot easier.

**Reply by mr-manuel** [reply]
I reworked a great part of the code today and documented it well. I also remade the [`battery_template.py`](https://github.com/mr-manuel/venus-os_dbus-serialbattery/blob/master/etc/dbus-serialbattery/bms/battery_template.py). If there are still questions open let me know.

**Reply by calledit** [reply]
Nice work! Will make it allot easier!

**First a question:**
What should I do with values that are required like self.max_battery_charge_current or self.max_battery_discharge_current which the LiTime BMS does not repport?

I could fill up the value for my specific battery but that is not a very good solution if other people want to use this and have different batteries.

**Then:**
I did notice that [this line](https://github.com/mr-manuel/venus-os_dbus-serialbattery/blob/e1215d58cd30b02d7ce9c3c7b3b7c38963902600/etc/dbus-serialbattery/bms/battery_template.py#L239C11-L239C38) is probably placed in the wrong function seams like it belongs to `def get_settings(self)`

And after scrolling through the above mentioned init function it seams like `self.total_ah_drawn` and `self.capacity_remain` is missing.

**Reply by mr-manuel** [reply]
I changed another ton of code and added a `History()` class. The new code should answer all your questions.

**Comment by traiannicolau** [comment]
I am also very interested on this, but I have 0 coding knowledge. How can I make this work? I have a raspberry pi zero w2 and 2 litime 12v 100ah group 24 batteries. Thanks!
